apps/spoc_info.1.md
% SPOC_INFO(1) SPOC User's Manual | Version 0
% spoc@zetamon.xyz
% December 25, 2021

# NAME

spoc_info - Print information about a SPOC file to stdout

# USAGE

spoc_info [*options*] < *filename.spoc*

# DESCRIPTION

Print header and/or summary information about a SPOC file

Information about each field in a SPOC file is given in the output. For
example, in non-json compact mode the field's range and quartiles are
listed.

Below, the 'x' field is shown to have a range of 249.950 meters. The
minimum value is 356631.450, and each quartile's size is shown following
the minimum value, and the last value is the field's maximum,
356881.400.

    x       range=249.950, 356631.450 +87.350 +45.920 +39.090 +77.590 = 356881.400

Below, the 'i' (intensity) field is shown to have a range of 4087. The
minimum value is 1, and each quartile's size is shown following the
minimum value, and the last value is the field's maximum, 4088.

This notation allows you to get an rough estimate of the shape of the
field's distribution. For example, notice how the size of the last
quartile is very large relative to first three quartiles. This indicates
that the intensity distribution is skewed toward zero. That is, most of
the intensity values (75% of them) are below 88 (1+19+35+33), while the
highest 25% of the values are spread out over the range 89 to 4088.

    i       range=4087, 1 +19 +35 +33 +4000 = 4088

# OPTIONS

--help, -h
:   Get help

--verbose, -v
:   Set verbose mode ON

--json, -j
: Toggle json output switch

--header-info, -e
: Toggle header information switch

--summary-info, -s
: Toggle summary information switch

--compact, -c
: Toggle compact output mode switch

# EXAMPLES

Compact, text mode output

    $ spoc_info < Austin.spoc
    major_version	0
    minor_version	1
    wkt	PROJCS["NAD83...]
    npoints	117456
    x	range=249.950, 356631.450 +87.350 +45.920 +39.090 +77.590 = 356881.400
    y	range=306.600, 3515084.930 +125.650 +52.810 +40.120 +88.020 = 3515391.530
    z	range=24.860, 1134.220 +2.290 +2.240 +3.290 +17.040 = 1159.080
    c	range=12, 1 +1 +0 +2 +9 = 13
    p	range=1, 152 +0 +1 +0 +0 = 153
    i	range=4087, 1 +19 +35 +33 +4000 = 4088

Compact, text mode output without header information

    $ spoc_info -h < Austin.spoc
    x	range=249.950, 356631.450 +87.350 +45.920 +39.090 +77.590 = 356881.400
    y	range=306.600, 3515084.930 +125.650 +52.810 +40.120 +88.020 = 3515391.530
    z	range=24.860, 1134.220 +2.290 +2.240 +3.290 +17.040 = 1159.080
    c	range=12, 1 +1 +0 +2 +9 = 13
    p	range=1, 152 +0 +1 +0 +0 = 153
    i	range=4087, 1 +19 +35 +33 +4000 = 4088

Non-compact, JSON output

    $ spoc_info -c -j < Austin.spoc
    {
        'header' : {
            'major_version' : 0,
            'minor_version' : 1,
            'npoints' : 117456,
            'wkt' : 'PROJCS["NAD83..."]]'
        },
        'summary' : {
            ...
            'c' : {
                'q0' : 1,
                'q1' : 2,
                'q2' : 2,
                'q3' : 4,
                'q4' : 13,
                'range' : 12
            },
            ...
            'x' : {
                'q0' : 356631.450000000011642,
                'q1' : 356718.800000000046566,
                'q2' : 356764.720000000030268,
                'q3' : 356803.810000000055879,
                'q4' : 356881.400000000023283,
                'range' : 249.950000000011642
            },
            'y' : {
                'q0' : 3515084.930000000167638,
                'q1' : 3515210.580000000074506,
                'q2' : 3515263.390000000130385,
                'q3' : 3515303.509999999776483,
                'q4' : 3515391.529999999795109,
                'range' : 306.599999999627471
            },
            'z' : {
                'q0' : 1134.220000000000027,
                'q1' : 1136.509999999999991,
                'q2' : 1138.750000000000000,
                'q3' : 1142.040000000000191,
                'q4' : 1159.080000000000155,
                'range' : 24.860000000000127
            }
        }
    }

# SEE ALSO

SPOC_SRS(1)


apps/spoc_info.cpp
#include "json.h"
#include "spoc.h"
#include "spoc_info_cmd.h"
#include "json.h"
#include <iostream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>

template<typename T,typename U>
spoc::json::object get_summary_object (const U &x)
{
    // Return value
    spoc::json::object s;

    // Handle the empty case
    if (x.empty ())
    {
        s["size"] = 0;
        s["min"] = 0;
        s["max"] = 0;
        return s;
    }

    // Copy
    auto y (x);

    // Get Quantiles
    const size_t n = y.size ();
    std::sort (begin (y), end (y));
    s["range"] = y.back() - y.front();
    s["q0"] = y.front();
    s["q1"] = y[n / 4];
    s["q2"] = y[n / 2];
    s["q3"] = y[3 * n / 4];
    s["q4"] = y.back();
    return s;
}

template<typename T,typename U>
std::string get_summary_string (const std::string &label, const U &x, const bool compact)
{
    // Return value
    std::stringstream s;
    s.precision (3);
    s << std::fixed;

    // Handle the empty case
    if (x.empty ())
    {
        if (compact)
        {
            return std::string ();
        }
        else
        {
            s << label << "0" << std::endl;
            s << "size\t0" << std::endl;
            s << "min\t0" << std::endl;
            s << "max\t0" << std::endl;
            return s.str ();
        }
    }

    // Copy
    auto y (x);

    // Get Quantiles
    const size_t n = y.size ();
    std::sort (begin (y), end (y));
    if (compact)
    {
        s << label;
        // Range
        s << "range=" << y.back() - y.front();
        // q0
        s << ", " << y.front();
        // Step to q1
        s << " +" << y[n / 4] - y.front ();
        // Step to q2
        s << " +" << y[n / 2] - y[n / 4];
        // Step to q3
        s << " +" << y[3 * n / 4] - y[n / 2];
        // Step to q4
        s << " +" << y.back () - y[3 * n / 4];
        // Q4
        s << " = " << y.back() << std::endl;
    }
    else
    {
        s << label << n << std::endl;
        s << "range\t" << y.back() - y.front() << std::endl;
        s << "q0\t" << y.front() << std::endl;
        s << "q1\t" << y[n / 4] << std::endl;
        s << "q2\t" << y[n / 2] << std::endl;
        s << "q3\t" << y[3 * n / 4] << std::endl;
        s << "q4\t" << y.back() << std::endl;
    }
    return s.str ();
}

// Process a spoc file and write to stdout
void process (const spoc::cmd::args &args, const spoc::spoc_file &f)
{
    using namespace std;
    using namespace spoc;

    cout.precision (15);
    cout << fixed;

    if (args.json)
    {
        json::object j;

        if (args.header_info)
        {
            json::object h;
            h["major_version"] = f.get_major_version ();
            h["minor_version"] = f.get_minor_version ();
            h["wkt"] = f.get_wkt ();
            h["npoints"] = f.get_npoints ();
            j["header"] = h;
        }

        if (args.summary_info)
        {
            json::object s;

            s["x"] = get_summary_object<double> (f.get_x ());
            s["y"] = get_summary_object<double> (f.get_y ());
            s["z"] = get_summary_object<double> (f.get_z ());
            s["c"] = get_summary_object<uint16_t> (f.get_c ());
            s["p"] = get_summary_object<uint16_t> (f.get_p ());
            s["i"] = get_summary_object<uint16_t> (f.get_i ());
            s["r"] = get_summary_object<uint16_t> (f.get_r ());
            s["g"] = get_summary_object<uint16_t> (f.get_g ());
            s["b"] = get_summary_object<uint16_t> (f.get_b ());
            json::array a;
            for (size_t k = 0; k < f.get_extra ().size (); ++k)
                a.push_back (get_summary_object<uint64_t> (f.get_extra ()[k]));
            s["extra"] = a;
            j["summary"] = s;
        }

        if (args.compact)
            spoc::json::operator<<(cout, j);
        else
            json::pretty_print (cout, j, 0);
    }
    else
    {
        if (args.header_info)
        {
            cout << "major_version\t" << int (f.get_major_version ()) << endl;
            cout << "minor_version\t" << int (f.get_minor_version ()) << endl;
            cout << "wkt\t" << f.get_wkt () << endl;
            cout << "npoints\t" << f.get_npoints () << endl;
        }

        if (args.summary_info)
        {
            cout << get_summary_string<double> ("x\t", f.get_x (), args.compact);
            cout << get_summary_string<double> ("y\t", f.get_y (), args.compact);
            cout << get_summary_string<double> ("z\t", f.get_z (), args.compact);
            cout << get_summary_string<uint16_t> ("c\t", f.get_c (), args.compact);
            cout << get_summary_string<uint16_t> ("p\t", f.get_p (), args.compact);
            cout << get_summary_string<uint16_t> ("i\t", f.get_i (), args.compact);
            cout << get_summary_string<uint16_t> ("r\t", f.get_r (), args.compact);
            cout << get_summary_string<uint16_t> ("g\t", f.get_g (), args.compact);
            cout << get_summary_string<uint16_t> ("b\t", f.get_b (), args.compact);
            for (size_t k = 0; k < f.get_extra ().size (); ++k)
            {
                stringstream s;
                s.precision (3);
                s << fixed;
                s << "extra " << k << "\t";
                cout << get_summary_string<uint16_t> (s.str (), f.get_extra ()[k], args.compact);
            }
        }
    }
}

int main (int argc, char **argv)
{
    using namespace std;
    using namespace spoc;

    try
    {
        // Parse command line
        const cmd::args args = cmd::get_args (argc, argv,
                string (argv[0]) + " [options] < spocfile");

        // If you are getting help, exit without an error
        if (args.help)
            return 0;

        // Show args
        if (args.verbose)
        {
            clog << "verbose\t" << args.verbose << endl;
            clog << "json\t" << args.json << endl;
            clog << "header-info\t" << args.header_info << endl;
            clog << "summary-info\t" << args.summary_info << endl;
            clog << "filenames\t" << args.fns.size () << endl;
        }

        if (args.fns.empty ())
        {
            if (args.verbose)
                clog << "Reading from stdin" << endl;

            // Read into spoc_file struct
            spoc_file f = read_spoc_file (cin);

            process (args, f);
        }
        else
        {
            for (auto fn : args.fns)
            {
                if (args.verbose)
                    clog << "Reading " << fn << endl;

                ifstream ifs (fn);

                if (!ifs)
                    throw runtime_error ("Could not open file for reading");

                // Read into spoc_file struct
                spoc_file f = read_spoc_file (ifs);

                process (args, f);
            }
        }

        return 0;
    }
    catch (const exception &e)
    {
        cerr << e.what () << endl;
        return -1;
    }
}
